monitoring with ngrok and datadog (#1272)
ngrok/ngrok-docs: docs/guides/other-guides/monitor-ngrok/monitor-ngrok.md

---
title: Observe your API traffic with ngrok and Datadog
sidebar_label: Monitor with Datadog
sidebar_position: 3
toc_max_heading_level: 3
---

## Introduction

This guide explains how to monitor your API or web app with ngrok by viewing traffic reports, using error request replays, and exporting logs and events to an external dashboard.

Whether you're an existing ngrok user looking to make your API more robust or a new user wondering whether ngrok meets your needs, this tutorial will demonstrate monitoring in detail.

## Prerequisites

> This guide assumes you already have [an ngrok account](https://dashboard.ngrok.com/signup) (a free account is sufficient) and have [Docker installed](https://docs.docker.com/get-started/get-docker) on your computer.

Reviewer: ```suggestion
This guide assumes you've already shipped ngrok as an API gateway using one of these guides:

- [Get started with ngrok's API gateway](https://ngrok.com/docs/guides/api-gateway/get-started/)
- [Deploy an API gateway with Kubernetes](https://ngrok.com/docs/guides/api-gateway/kubernetes/)
- [Deploy a multicloud API gateway](https://ngrok.com/docs/guides/api-gateway/multicloud/)
```

:::note
The API uses a named network, `ngrokTest`, so that in the next section, you can start the ngrok Agent on the same network as the API.
:::

![Simple web API](./img/simple-web-api.png)

## Publish your API with ngrok

Now that the API is running locally, you can expose it on a public URL, using the [Docker image for the ngrok agent](https://hub.docker.com/r/ngrok/ngrok).

- Replace the authentication token in the command below with your token from the [ngrok dashboard](https://dashboard.ngrok.com/get-started/your-authtoken), and run the command in a new terminal.

  ```sh
  docker run -it --rm --platform=linux/amd64 --network=ngrokTest -e NGROK_AUTHTOKEN=Y0urS3cr3tK3y ngrok/ngrok:3.22.0-alpine-amd64 http http://api
  ```

  This command starts the ngrok agent locally and connects it to the API running on the `api` container.

- Browse to the URL labeled `Forwarding`, which should look like this in your terminal: `https://eb45-79-127-145-72.ngrok-free.app`.
- You can now see your request going from the browser to the ngrok agent you're running in one terminal window and then to the API in your other terminal window.

> ![Published API](./img/api-gent.png)

Reviewer: What I'd prefer to do here is cut all this and instead use the change I requested above, which points folks to the previous API gateway guides, so that they already have the API demo app running and their whole API gateway configured.

## Monitor your API

You can monitor web servers in two ways in ngrok, with the Traffic Inspector and with [Events](https://ngrok.com/docs/obs/events).

The Traffic Inspector is a filterable list of your API's requests and responses, available on the ngrok dashboard. The Traffic Inspector is useful for viewing error details and for replaying requests to test new policies and bug fixes. The inspector is a manual way to monitor your API. Request data is kept for three days (or for 90 days as a paid extra).

An event is the data that ngrok provides about a request, which is exported to a dedicated monitoring platform. Events offer an automated means of monitoring your API. Events are also the only way for your team to get automatic error alerts (notifications) instead of constantly having to check for errors on a dashboard.

At the time of writing, ngrok allows you to export events only to AWS, Azure, and Datadog. We do not support event exports to OpenTelemetry or custom URLs, such as self-hosted servers, so you have to use a paid cloud service. You can't perform custom processing or use Elastic, Prometheus, Splunk, or alternative monitoring apps.

> ## Use the Traffic Inspector

Reviewer: ```suggestion
## Observe API traffic with Traffic Inspector
```

- Click a request in the Traffic Inspector list, then click **Enable full capture** in the sidebar.

  ![Enable full capture](./img/enableFullCapture.png)

- This button takes you to your account settings, where you can enable full capture under **Observability**.

  ![Enable full capture settings](./img/enableFullCaptureAccount.png)

- Then, return to your published API URL and refresh the browser page a few times to send fresh requests through ngrok.
- In the ngrok Traffic Inspector, click on the event at the top of the list.

  The event should now show all the request and response details and content. If the full capture details don't appear, you may need to restart your ngrok agent in the terminal.

- Click the **Replay** button and notice that the request is resent from the ngrok host.

  ![Replay request](./img/replay.png)

- You can also replay a request with changes to alter any of the headers or POST data.

  ![Replay with changes](./img/replayChanges2.png)

Replaying requests is useful for debugging. For example, you could find the request that caused an error in your API, deploy a fix for the API, and replay the request to confirm you've fixed the issue.

> ### Traffic Policy example

Reviewer: ```suggestion
### Observe and debug Traffic Policy rules with request replay
```

- Stop the ngrok Docker container in the terminal with <kbd>Ctrl</kbd><kbd>+</kbd><kbd>C</kbd>.
- Create a file named `policy.yml` with the code below.

  ```yml
  on_http_request:
    - actions:
      - type: rate-limit
        config:
          name: 1 request per minute
          algorithm: sliding_window
          capacity: 1
          rate: 1m
          bucket_key:
            - "hasReqHeader('host') ? getReqHeader('host')[0] : 'unknown'"
  ```

- Run the agent again with the command below. Replace the authentication token and the URL with your token and URL (using `https`, not `http`).

  ```sh
  docker run -it --rm --platform=linux/amd64 --network=ngrokTest -v ".:/app" -w "/app" -e NGROK_AUTHTOKEN=Y0urS3cr3t ngrok/ngrok:3.22.0-alpine-amd64 http http://api --traffic-policy-file policy.yml --url https://massive-pelican-arguably.ngrok-free.app
  ```

- Browse to the API URL and refresh the page a few times. Notice that the page stops responding.
- On the ngrok Traffic Inspector page, note that the error code **`429`** (rate limit) was returned and that the duration of the request was instant and caused no load on your API.

  ![Rate limit](./img/rateLimit.png)

> - Edit the `policy.yml` file and change `capacity` to `10`, then restart the ngrok container.

Reviewer: Given that we're funneling folks through the API gateway guides first, we can remove these steps and just direct folks to see the impact of the rate limiting Traffic Policy action they've already set up.

- In the Traffic Inspector, click on one of the **`429`** events, then click **Replay**. Note that the request now responds without error because the rate limit has been increased.

  ![Replay after rate limit adjustment](./img/replayLimit.png)

> ## Monitor events

Reviewer: ```suggestion
## Monitor API gateway traffic with Datadog
```

In this section, you'll learn how to export ngrok events to the Datadog monitoring service.

There are two [types of events](https://ngrok.com/docs/obs/events/reference): standard traffic events (requests to your API) and audit events (changes to secret keys and URLs). For this simple example, you'll monitor a traffic event.

Before adding an event subscription, you need somewhere to send events:

- Sign up for a [Datadog](https://www.datadoghq.com) trial account. (You will eventually need a paid Datadog plan to continue using logs and monitors.)

>   You can enter anonymous dummy values instead of your personal information for all the Datadog registration fields except your email address, which you need to confirm. You also can't skip the third step of the sign-up process, in which you create a Datadog agent somewhere.

Reviewer: I don't love this tip of using dummy values—feels incongruous with our target ICP here, who are more DevOps/infra engineers with production API gateways.

- In Step 3 of the Datadog sign-up, click on **Docker** in the sidebar. Copy and paste the given command into a terminal and run it.

  ![Join Datadog](./img/joinDatadog.png)

- Once the Datadog site notices your agent is running, click **Finish** at the bottom of the page.
- You now no longer need the agent running locally and can remove the Datadog container with the command below.

  ```sh
  docker stop dd-agent
  docker remove dd-agent
  docker image rm gcr.io/datadoghq/agent:7

>   ```

Reviewer: Can you get an API key this way? https://docs.datadoghq.com/account_management/api-app-keys/#add-an-api-key-or-client-token

- Take note of your Datadog API key, which you can access by clicking your username at the bottom left of the sidebar and clicking **API Keys** in the menu that opens.

  ![Datadog API key](./img/datadogKey.png)

- In the ngrok navigation panel, browse to the [**Events Stream**](https://dashboard.ngrok.com/event-subscriptions) and add a new subscription.

  ![Add event subscription](./img/makeSubscription.png)

- In the **New Event Subscription** sidebar, enter `traffic` as the **Description** and then **Add** a new source (event type). Choose **`http_request_complete`**.

  ![Add source](./img/addEventType.png)

- Add Datadog as a destination, add the Datadog site and API key you noted earlier, and send a test event.

- Open the Datadog site and browse to the **Log Explorer** page from the navigation panel. Enable logs. You should see that the event from ngrok has appeared.

  ![Datadog logs](./img/datadogLog.png)

- In ngrok, click **Done** and **Save**.

- Refresh your ngrok API page a few times so that new requests are logged in Datadog.

> ## Create a dashboard

Reviewer: ```suggestion
### Create a dashboard
```

  ![Datadog default dashboard](./img/datadogDefaultDashboard.png)

- In the cloned dashboard, click **Add widgets**.

  ![Datadog add widget](./img/datadogAddWidget.png)

Configure the widget as follows.

- **Step 1:** Choose **Query value** as your visualization.
- **Step 2:** Change **Metrics** to **Logs** (ngrok does not export metrics) and enter the log filter, `@http.status_code:500`, so that Datadog only counts errors.
- **Step 3:** Set the time preference to **Past 1 Hour**.
- **Step 4:** Name the widget `Errors in the last hour`.

  ![Datadog edit widget](./img/datadogWidgetDetails.png)

Your new widget will be available in the dashboard, allowing your support staff to see instantly whether any errors have occurred in your API.

![Datadog custom dashboard](./img/datadogCustomDashboard.png)

Since the sample API never returns errors, an easy way to test the `Errors in the last hour` widget is to stop the sample API Docker container and then try to browse to the site on the public ngrok endpoint.

If you want to create widgets for other log information, you can see which fields are available by reading the JSON of any event you click on in the log inspector.

> ## Create a notification

Reviewer: ```suggestion
### Create a notification
```

## Further reading

To learn more about any of the concepts in this guide, consult the following articles:

- The [Get Docker](https://docs.docker.com/get-started/get-docker) guide
- The guide to [getting started with Docker for ngrok](https://dashboard.ngrok.com/get-started/setup/docker)
- The complete [Docker ngrok agent](https://ngrok.com/docs/using-ngrok-with/docker) documentation
- The [ngrok Traffic Policy](https://ngrok.com/docs/traffic-policy) documentation
- The [ngrok Traffic Policy actions](https://ngrok.com/docs/traffic-policy/actions/rate-limit) documentation
- The [ngrok Cloud Endpoints](https://ngrok.com/docs/universal-gateway/cloud-endpoints) documentation
- The [ngrok Agent Endpoints](https://ngrok.com/docs/universal-gateway/agent-endpoints) documentation
- The [ngrok Events](https://ngrok.com/docs/obs/events) documentation
- [Datadog](https://www.datadoghq.com)

> - [ntfy](https://ntfy.sh)

Reviewer: Not sure why this link is here.
